src/tokenizedSubaccounts/PPTSA.sol

```diff
@@ -0,0 +1,559 @@
+      revert PPT_InvalidParams();
+    }
+
+    if (tradeData.trades[0].asset == address(tsaAddresses.wrappedDepositAsset)) {
+      if (tradeData.trades.length != 1) {
+        revert PPT_InvalidParams();
+      }
+      if (tradeData.trades[0].asset != address(tsaAddresses.wrappedDepositAsset)) {
+        revert PPT_InvalidAsset();
+      }
+      if (tradeData.trades[0].amount > 0) {
+        _verifyCollateralBuy(tradeData, tsaAddresses);
+      } else {
+        _verifyCollateralSell(tradeData, tsaAddresses);
+      }
+      return;
+    } else if (tradeData.trades[0].asset == address(_getPPTSAStorage().optionAsset)) {
+      _verifyRfqExecute(tradeData);
+    } else {
+      revert PPT_InvalidAsset();
+    }
+  }
+
+  function _verifyCollateralBuy(IRfqModule.RfqOrder memory tradeData, BaseTSAAddresses memory tsaAddresses)
+    internal
+    view
+  {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+
+    int cashBalance = tsaAddresses.subAccounts.getBalance(subAccount(), tsaAddresses.cash, 0);
+    if (cashBalance <= 0) {
+      revert PPT_MustHavePositiveCash();
+    }
+    uint basePrice = _getBasePrice();
+
+    _verifyFee(tradeData.maxFee, basePrice);
+
+    if (tradeData.trades[0].price > basePrice.multiplyDecimal($.ppParams.worstSpotBuyPrice)) {
+      revert PPT_SpotLimitPriceTooHigh();
+    }
+
+    int cost = tradeData.trades[0].price.toInt256().multiplyDecimal(tradeData.trades[0].amount);
+    int bufferedBalance = cashBalance.multiplyDecimal($.ppParams.spotTransactionLeniency);
+    if (cost > bufferedBalance) {
+      revert PPT_BuyingTooMuchCollateral();
+    }
+  }
+
+  function _verifyCollateralSell(IRfqModule.RfqOrder memory tradeData, BaseTSAAddresses memory tsaAddresses)
+    internal
+    view
+  {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+
+    int cashBalance = tsaAddresses.subAccounts.getBalance(subAccount(), tsaAddresses.cash, 0);
+    if (cashBalance >= 0) {
+      revert PPT_MustHaveNegativeCash();
+    }
+
+    uint basePrice = _getBasePrice();
+
+    _verifyFee(tradeData.maxFee, basePrice);
+
+    if (tradeData.trades[0].price < basePrice.multiplyDecimal($.ppParams.worstSpotSellPrice)) {
+      revert PPT_SpotLimitPriceTooLow();
+    }
+
+    // cost is positive, balance is negative
+    uint cost = tradeData.trades[0].price.multiplyDecimal(tradeData.trades[0].amount.toUint256());
+    int bufferedBalance = cashBalance.multiplyDecimal($.ppParams.spotTransactionLeniency);
+
+    // We make sure we're not selling more $ value of collateral than we have in debt
+    if (cost > bufferedBalance.abs()) {
+      revert PPT_SellingTooMuchCollateral();
+    }
+  }
+
+  function _verifyRfqExecute(IRfqModule.RfqOrder memory tradeData) internal view {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+    if (tradeData.trades.length != 2) {
+      revert PPT_InvalidParams();
+    }
+    if (tradeData.trades[0].asset != tradeData.trades[1].asset) {
+      revert PPT_InvalidParams();
+    }
```

Comment:
Nice checks, easy to overlook 👍 Would've merged into one if statement to save on contract size though.

Reply:
Sounds good. Sorry, contract size is a new thing Im getting used to. Thanks for pointing these out.